File: daemon/sdk/project_sync.go
package mindwire

import (
	"context"
	"errors"
	"fmt"

	"github.com/oblien/mindwire/daemon/internal/conversations"
	"github.com/oblien/mindwire/daemon/internal/projectsync"
	"github.com/oblien/mindwire/daemon/internal/registry"
)

type ProjectSyncRequest = projectsync.Request
type ProjectSyncOperation = projectsync.Operation
type ProjectSyncCheckpoint = projectsync.Checkpoint
type ProjectSyncConflict = projectsync.Conflict
type ProjectSyncObjectPage = projectsync.ObjectPage
type ProjectSync struct{ c *Client }

func (co *core) Busy(id string) bool { return co.sup.Busy(id) }
func (co *core) BusyPath(path string) bool {
	return co.sup.BusyPath(path) || co.execution.BusyPath(path)
}

func (co *core) discoverForSync(ctx context.Context, id string) error {
	issues, err := co.conversations.Refresh(ctx, conversations.Query{ProjectID: id, Refresh: true})
	if err != nil {
		return err
	}
	if len(issues) > 0 {
		return fmt.Errorf("refresh native conversations before synchronizing: %s", issues[0].Message)
	}
	return nil
}
func (s *ProjectSync) Export(req ProjectSyncRequest) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Start("export", req)
	return v, workspaceError("Workspace.Sync.Export", e)
}
func (s *ProjectSync) Import(req ProjectSyncRequest) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Start("import", req)
	return v, workspaceError("Workspace.Sync.Import", e)
}
func (s *ProjectSync) Get(id string) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Get(id)
	return v, workspaceError("Workspace.Sync.Get", e)
}
func (s *ProjectSync) List() ([]ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.List()
	return v, workspaceError("Workspace.Sync.List", e)
}
func (s *ProjectSync) Retry(id string) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Retry(id)
	return v, workspaceError("Workspace.Sync.Retry", e)
}
func (s *ProjectSync) Cancel(id string) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Cancel(id)
	return v, workspaceError("Workspace.Sync.Cancel", e)
}
func (s *ProjectSync) Wait(ctx context.Context, id string) (ProjectSyncOperation, error) {
	v, e := s.c.core.projectSync.Wait(ctx, id)
	return v, workspaceError("Workspace.Sync.Wait", e)
}
func (s *ProjectSync) Checkpoint(id string) (ProjectSyncCheckpoint, error) {
	v, e := s.c.core.projectSync.Checkpoint(id)
	return v, workspaceError("Workspace.Sync.Checkpoint", e)
}
func (s *ProjectSync) AcceptCheckpoint(ctx context.Context, c ProjectSyncCheckpoint) error {
	return workspaceError("Workspace.Sync.AcceptCheckpoint", s.c.core.projectSync.AcceptCheckpoint(ctx, c))
}
func (s *ProjectSync) Objects(id string, cursor int) (ProjectSyncObjectPage, error) {
	v, e := s.c.core.projectSync.Objects(id, cursor)
	return v, workspaceError("Workspace.Sync.Objects", e)
}
func (s *ProjectSync) Missing(ids []string) ([]string, error) {
	v, e := s.c.core.projectSync.Missing(ids)
	return v, workspaceError("Workspace.Sync.Missing", e)
}
func (s *ProjectSync) GetObject(id string) ([]byte, error) {
	v, e := s.c.core.projectSync.GetObject(id)
	return v, workspaceError("Workspace.Sync.GetObject", e)
}
func (s *ProjectSync) PutObject(id string, data []byte) error {
	return workspaceError("Workspace.Sync.PutObject", s.c.core.projectSync.PutObject(id, data))
}

// RelayCheckpoint transfers only missing immutable chunks, parents first. The
// caller may retry after losing connectivity; verified objects are reused.
func (s *ProjectSync) RelayCheckpoint(ctx context.Context, target *ProjectSync, id string, progress func(int64)) error {
	if _, err := target.Checkpoint(id); err == nil {
		return nil
	} else if !errors.Is(err, registry.ErrNotFound) {
		return err
	}
	c, err := s.Checkpoint(id)
	if err != nil {
		return err
	}
	for _, p := range c.Parents {
		if err = s.RelayCheckpoint(ctx, target, p, progress); err != nil {
			return err
		}
	}
	for cursor := 0; ; {
		if err = ctx.Err(); err != nil {
			return err
		}
		page, err := s.Objects(id, cursor)
		if err != nil {
			return err
		}
		missing, err := target.Missing(page.Objects)
		if err != nil {
			return err
		}
		for _, object := range missing {
			if err = ctx.Err(); err != nil {
				return err
			}
			b, err := s.GetObject(object)
			if err != nil {
				return err
			}
			if err = target.PutObject(object, b); err != nil {
				return err
			}
			if progress != nil {
				progress(int64(len(b)))
			}
		}
		if page.Next == 0 {
			break
		}
		cursor = page.Next
	}
	return target.AcceptCheckpoint(ctx, c)
}

// SwitchTo preserves the source and opens no chat or process. Stable request IDs
// make a lost acknowledgement safe. A conflict is returned as an operation so
// callers can show its paths and retained recovery checkpoint instead of hiding it.
func (s *ProjectSync) SwitchTo(ctx context.Context, target *ProjectSync, req ProjectSyncRequest, progress func(int64)) (ProjectSyncOperation, error) {
	if s.c.core == target.c.core {
		return ProjectSyncOperation{}, workspaceError("Workspace.Sync.SwitchTo", registry.ErrInvalid)
	}
	if accepted, err := target.Get(req.ID + "-import"); err == nil {
		if accepted.Path != req.Path {
			return accepted, workspaceError("Workspace.Sync.SwitchTo", registry.ErrConflict)
		}
		return target.Wait(ctx, accepted.ID)
	} else if !errors.Is(err, registry.ErrNotFound) {
		return accepted, err
	}
	o, err := s.Export(ProjectSyncRequest{ID: req.ID + "-export", ProjectID: req.ProjectID})
	if err != nil {
		return o, err
	}
	o, err = s.Wait(ctx, o.ID)
	if err != nil || o.Status != "succeeded" {
		return o, err
	}
	if err = s.RelayCheckpoint(ctx, target, o.ResultCheckpointID, progress); err != nil {
		return o, err
	}
	o, err = target.Import(ProjectSyncRequest{ID: req.ID + "-import", CheckpointID: o.ResultCheckpointID, Path: req.Path})
	if err != nil {
		return o, err
	}
	return target.Wait(ctx, o.ID)
}

File: packages/sdk/src/project-sync.ts
import type { Mindwire } from "./client.js";
import { ApiError, MindwireError } from "./errors.js";

export interface ProjectSyncEntry {
  kind: "file" | "directory" | "symlink";
  mode: number;
  size: number;
  chunks?: string[];
  link?: string;
  jsonLines?: boolean;
}

export interface ProjectSyncCheckpoint {
  id: string;
  projectId: string;
  parents: string[];
  manifest: ProjectSyncEntry;
  objectCount: number;
  bytes: number;
}

export interface ProjectSyncRequest {
  /** Durable idempotency key. Keep the same ID after a lost acknowledgement. */
  id: string;
  projectId?: string;
  checkpointId?: string;
  /** Only needed on first import; subsequent imports find the preserved replica. */
  path?: string;
}

export interface ProjectSyncConflict {
  path: string;
  kind: "file" | "conversation" | "git" | "recovery";
  message: string;
}

export interface ProjectSyncOperation extends ProjectSyncRequest {
  kind: "export" | "import";
  status: "queued" | "exporting" | "preparing" | "applying" | "succeeded" | "conflicts" | "failed" | "recovery_required";
  phase: string;
  progress: number;
  resultProjectId?: string;
  resultCheckpointId?: string;
  recoveryCheckpointId?: string;
  conflicts?: ProjectSyncConflict[];
  error?: string;
  createdAt: string;
  updatedAt: string;
}

export interface ProjectSyncProgress {
  phase: "checkpoint" | "transfer" | "apply";
  transferredBytes: number;
  operation?: ProjectSyncOperation;
}

export class ProjectSyncError extends MindwireError {
  constructor(readonly operation: ProjectSyncOperation) {
    super(operation.error ?? (operation.status === "conflicts"
      ? "Changes conflict. Both workspace copies and their checkpoints are preserved."
      : `Project synchronization ${operation.status}.`));
    this.name = "ProjectSyncError";
  }
}

const active = (o: ProjectSyncOperation): boolean => ["queued", "exporting", "preparing", "applying"].includes(o.status);
const prefix = "/workspace/sync";
type Options = { signal?: AbortSignal };

function delay(ms: number, signal?: AbortSignal): Promise<void> {
  signal?.throwIfAborted();
  return new Promise((resolve, reject) => {
    const done = () => { signal?.removeEventListener("abort", abort); resolve(); };
    const timer = setTimeout(done, ms);
    const abort = () => { clearTimeout(timer); signal?.removeEventListener("abort", abort); reject(signal?.reason); };
    signal?.addEventListener("abort", abort, { once: true });
  });
}

/** Shared checkpoint protocol over the existing authenticated workspace transport.
 * No Git hosting service or third-party storage is involved. Native adapters in
 * the daemon own export/import; clients only relay verified immutable chunks.
 */
export class ProjectSyncApi {
  private readonly flights = new Map<string, { intent: string; target: ProjectSyncApi; result: Promise<ProjectSyncOperation> }>();
  constructor(private readonly mw: Mindwire) {}

  export(request: ProjectSyncRequest, options: Options = {}): Promise<ProjectSyncOperation> {
    return this.mw.http.request("POST", `${prefix}/exports`, { body: request, ...options });
  }
  import(request: ProjectSyncRequest, options: Options = {}): Promise<ProjectSyncOperation> {
    return this.mw.http.request("POST", `${prefix}/imports`, { body: request, ...options });
  }
  async list(options: Options = {}): Promise<ProjectSyncOperation[]> {
    const response = await this.mw.http.request<{ operations: ProjectSyncOperation[] }>("GET", `${prefix}/operations`, options);
    return response.operations;
  }
  get(id: string, options: Options = {}): Promise<ProjectSyncOperation> {
    return this.mw.http.request("GET", `${prefix}/operations/${encodeURIComponent(id)}`, options);
  }
  retry(id: string, options: Options = {}): Promise<ProjectSyncOperation> {
    return this.mw.http.request("POST", `${prefix}/operations/${encodeURIComponent(id)}/retry`, options);
  }
  cancel(id: string): Promise<ProjectSyncOperation> {
    return this.mw.http.request("POST", `${prefix}/operations/${encodeURIComponent(id)}/cancel`);
  }
  checkpoint(id: string, options: Options = {}): Promise<ProjectSyncCheckpoint> {
    return this.mw.http.request("GET", `${prefix}/checkpoints/${encodeURIComponent(id)}`, options);
  }
  acceptCheckpoint(checkpoint: ProjectSyncCheckpoint, options: Options = {}): Promise<ProjectSyncCheckpoint> {
    return this.mw.http.request("PUT", `${prefix}/checkpoints/${encodeURIComponent(checkpoint.id)}`, { body: checkpoint, ...options });
  }
  objects(id: string, cursor = 0, options: Options = {}): Promise<{ objects: string[]; next?: number }> {
    return this.mw.http.request("GET", `${prefix}/checkpoints/${encodeURIComponent(id)}/objects`, { query: { cursor }, ...options });
  }
  async missing(objects: string[], options: Options = {}): Promise<string[]> {
    const response = await this.mw.http.request<{ objects: string[] }>("POST", `${prefix}/objects/missing`, { body: { objects }, ...options });
    return response.objects;
  }
  /** Base64 chunk, at most 256 KiB before encoding. */
  async getObject(id: string, options: Options = {}): Promise<string> {
    const response = await this.mw.http.request<{ data: string }>("GET", `${prefix}/objects/${encodeURIComponent(id)}`, options);
    return response.data;
  }
  async putObject(id: string, data: string, options: Options = {}): Promise<void> {
    await this.mw.http.request("PUT", `${prefix}/objects/${encodeURIComponent(id)}`, { body: { data }, ...options });
  }
  async wait(id: string, options: Options & { onUpdate?: (operation: ProjectSyncOperation) => void } = {}): Promise<ProjectSyncOperation> {
    for (;;) {
      const o = await this.get(id, { signal: options.signal });
      options.onUpdate?.(o);
      if (!active(o)) return o;
      await delay(500, options.signal);
    }
  }

  /** Repeating this relay skips acknowledged checkpoints and verified chunks.
   * Aborting detaches the relay; daemon export/apply operations remain durable.
   */
  async relayCheckpoint(target: ProjectSyncApi, id: string, options: Options & { onBytes?: (bytes: number) => void } = {}): Promise<void> {
    const visited = new Set<string>();
    const visit = async (checkpointId: string): Promise<void> => {
      options.signal?.throwIfAborted();
      if (visited.has(checkpointId)) return;
      visited.add(checkpointId);
      try { await target.checkpoint(checkpointId, options); return; }
      catch (error) { if (!(error instanceof ApiError) || error.status !== 404) throw error; }
      const checkpoint = await this.checkpoint(checkpointId, options);
      for (const parent of checkpoint.parents) await visit(parent);
      let cursor = 0;
      do {
        const page = await this.objects(checkpointId, cursor, options);
        const missing = await target.missing(page.objects, options);
        // Bounded chunk streaming: no archive/project-sized RAM buffer.
        for (const object of missing) {
          const data = await this.getObject(object, options);
          await target.putObject(object, data, options);
          const padding = data.endsWith("==") ? 2 : data.endsWith("=") ? 1 : 0;
          options.onBytes?.(Math.floor(data.length * 3 / 4) - padding);
        }
        cursor = page.next ?? 0;
      } while (cursor);
      await target.acceptCheckpoint(checkpoint, options);
    };
    await visit(id);
  }

  /** Safe X → Y and Y → X. Persist request.id until acknowledged. A conflict
   * throws ProjectSyncError carrying the complete conflict/recovery operation.
   */
  switchTo(target: ProjectSyncApi, request: { id: string; projectId: string; path?: string },
    options: Options & { onProgress?: (progress: ProjectSyncProgress) => void } = {}): Promise<ProjectSyncOperation> {
    const intent = JSON.stringify(request);
    const old = this.flights.get(request.id);
    if (old) {
      if (old.intent !== intent || old.target !== target) return Promise.reject(new MindwireError("This sync ID already belongs to a different request or destination."));
      return old.result;
    }
    const work = async (): Promise<ProjectSyncOperation> => {
      if (target === this || target.mw.http === this.mw.http) throw new MindwireError("Choose another workspace for this project.");
      // Once the destination accepted the immutable checkpoint, reconnecting
      // only needs that destination. The original laptop may already be offline.
      let accepted: ProjectSyncOperation | undefined;
      try { accepted = await target.get(`${request.id}-import`, options); }
      catch (error) { if (!(error instanceof ApiError) || error.status !== 404) throw error; }
      if (accepted) {
        if ((accepted.path ?? "") !== (request.path ?? "")) throw new MindwireError("This sync ID belongs to a different destination folder.");
        const operation = await target.wait(accepted.id, { signal: options.signal,
          onUpdate: operation => options.onProgress?.({ phase: "apply", transferredBytes: 0, operation }) });
        if (operation.status !== "succeeded") throw new ProjectSyncError(operation);
        return operation;
      }
      const health = await Promise.all([this.mw.health(), target.mw.health()]);
      if (health.some(h => (h.projectSyncVersion ?? 0) < 1)) throw new MindwireError("Update Mindwire on both workspaces to switch this project.");
      let bytes = 0;
      const emit = (phase: ProjectSyncProgress["phase"], operation?: ProjectSyncOperation) => options.onProgress?.({ phase, transferredBytes: bytes, operation });
      emit("checkpoint");
      let o = await this.export({ id: `${request.id}-export`, projectId: request.projectId }, options);
      o = await this.wait(o.id, { signal: options.signal, onUpdate: value => emit("checkpoint", value) });
      if (o.status !== "succeeded" || !o.resultCheckpointId) throw new ProjectSyncError(o);
      emit("transfer");
      await this.relayCheckpoint(target, o.resultCheckpointId, { signal: options.signal, onBytes: n => { bytes += n; emit("transfer"); } });
      o = await target.import({ id: `${request.id}-import`, checkpointId: o.resultCheckpointId, path: request.path }, options);
      o = await target.wait(o.id, { signal: options.signal, onUpdate: value => emit("apply", value) });
      if (o.status !== "succeeded") throw new ProjectSyncError(o);
      return o;
    };
    const result = work().finally(() => this.flights.delete(request.id));
    this.flights.set(request.id, { intent, target, result });
    return result;
  }
}
